## 4.9: Intro to Data Visualization with Python - Part 1
### 1. Import libraries
### 2. Import Data Frame
### 3. Data Wrangling
### 4. Data Consistency Check
### 5. Export Data Frame


## 1. Import Libraries

In [3]:
## import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

## 2. Import Data Frame

In [4]:
## import data frame

path = r'/Users/eriseldabaci/Desktop/CareerFoundry/Python Fundamentals for Data Analysts/Instacart Basket Analysis'

In [5]:
path

'/Users/eriseldabaci/Desktop/CareerFoundry/Python Fundamentals for Data Analysts/Instacart Basket Analysis'

In [6]:
df_customers = pd.read_csv(os.path.join(path,'02 Data','Original Data','customers.csv'))

## 3. Data Wrangling

In [7]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [8]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
df_customers.shape

(206209, 10)

In [10]:
## Dropping Columns - 'fam_status' in our anaylisis. It isn't necessary for or relevant to our analysis.

df_customers = df_customers.drop(columns=['fam_status'])

In [11]:
## Renaming Column 'Surnam' and 'STATE', 'Income', 'Fisrt Name'.

df_customers.rename(columns = {'Surnam' : 'Surname'}, inplace = True)
df_customers.rename(columns = {'STATE' : 'State'}, inplace = True)
df_customers.rename(columns = {'income' : 'Income'}, inplace = True)


In [12]:
df_customers.rename(columns = {'First Name' : 'Name'}, inplace = True)

In [13]:
df_customers.head()

user_id      Name   Surname  Gender       State  Age date_joined  \
0    26711   Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890  Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803   Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935  Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797       Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants  Income  
0             3  165665  
1             0   59285  
2             2   99568  
3             0   42049  
4             1   40374

In [14]:
## Check the type for columns into Data Frame

df_customers.dtypes

user_id          int64
Name            object
Surname         object
Gender          object
State           object
Age              int64
date_joined     object
n_dependants     int64
Income           int64
dtype: object

In [15]:
## Changing a Variable’s Data Type

df_customers['user_id'] = df_customers['user_id'].astype('str')

## 3. Data Consistency Check


In [16]:
## Missing Values

df_customers.isnull().sum()

user_id             0
Name            11259
Surname             0
Gender              0
State               0
Age                 0
date_joined         0
n_dependants        0
Income              0
dtype: int64

### Into column 'Name' missing value are 11259 and we can leave as they are because don't have any impact during analysis.

In [17]:
## Mixed Data Type

for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

Name


In [18]:
## Change column " First Name" from mixed tye to string

df_customers['Name'] = df_customers['Name'].astype(str)

In [19]:
## Check 'First Name' column type

df_customers['Name'].dtype

dtype('O')

## 4. Merge Data Set

In [20]:
## Import Data Frame

df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data ', 'orders_products_combined_update2.pkl'))

In [21]:
## check data frame 'orders'

df_ords_prods.shape

(32404859, 26)

In [22]:
## See the first 5 rows

df_ords_prods.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True       196.0                1.0   
1                    15.0         False       196.0                1.0   
2                    21.0         False       196.0                1.0   
3                    29.0         False       196.0                1.0   
4                    28.0         False       196.0                1.0   

   reordered  ...     Busiest day  Busyness_Level    Busiest days  \
0        0.0  ...  Regularly busy  Regularly busy  Regularly busy   
1        1.0  ...  Regularly busy    Slowest days    Slowest days   
2        1.0  ...  Regularly busy    Slowest days    Slowest days   
3        1.0  ...      Least busy    Slowest days    Slowest days   
4        1.0  ...      Least busy    Slowest days    Slowest days   

   busiest_period_of_day  max_order  loyalty_flag average_price spending_flag  \
0          Fewest orders         10  New customer      6.367797   Low spender   
1          Fewest orders         10  New customer      6.367797   Low spender   
2         Average orders         10  New customer      6.367797   Low spender   
3          Fewest orders         10  New customer      6.367797   Low spender   
4          Fewest orders         10  New customer      6.367797   Low spender   

  ordering_behavior        order_frequency  
0              20.5  Non-frequent customer  
1              20.5  Non-frequent customer  
2              20.5  Non-frequent customer  
3              20.5  Non-frequent customer  
4              20.5  Non-frequent customer  

[5 rows x 26 columns]

In [23]:
## check data frame 'customers'

df_customers.shape

(206209, 9)

In [24]:
## See the first 5 rows

df_customers.head()

user_id      Name   Surname  Gender       State  Age date_joined  \
0   26711   Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890  Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803   Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935  Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797       Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants  Income  
0             3  165665  
1             0   59285  
2             2   99568  
3             0   42049  
4             1   40374

In [29]:
## The user_id the same data type and after that merge the data frames

df_ords_prods['user_id'] = df_ords_prods['user_id'].astype(str)
df_customers['user_id'] = df_customers['user_id'].astype(str)

df_merged = pd.merge(df_ords_prods, df_customers, on='user_id')

In [30]:
df_merged.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True       196.0                1.0   
1                    15.0         False       196.0                1.0   
2                    21.0         False       196.0                1.0   
3                    29.0         False       196.0                1.0   
4                    28.0         False       196.0                1.0   

   reordered  ... ordering_behavior        order_frequency   Name  Surname  \
0        0.0  ...              20.5  Non-frequent customer  Linda   Nguyen   
1        1.0  ...              20.5  Non-frequent customer  Linda   Nguyen   
2        1.0  ...              20.5  Non-frequent customer  Linda   Nguyen   
3        1.0  ...              20.5  Non-frequent customer  Linda   Nguyen   
4        1.0  ...              20.5  Non-frequent customer  Linda   Nguyen   

   Gender    State Age date_joined n_dependants Income  
0  Female  Alabama  31   2/17/2019            3  40423  
1  Female  Alabama  31   2/17/2019            3  40423  
2  Female  Alabama  31   2/17/2019            3  40423  
3  Female  Alabama  31   2/17/2019            3  40423  
4  Female  Alabama  31   2/17/2019            3  40423  

[5 rows x 34 columns]

In [33]:
df_ords_prods.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'new_customer',
       'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'product_name', 'aisle_id', 'department_id', 'prices',
       'price_range_loc', 'Busiest day', 'Busyness_Level', 'Busiest days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag', 'average_price',
       'spending_flag', 'ordering_behavior', 'order_frequency'],
      dtype='object')

In [34]:
df_customers.columns

Index(['user_id', 'Name', 'Surname', 'Gender', 'State', 'Age', 'date_joined',
       'n_dependants', 'Income'],
      dtype='object')

In [35]:
df_merged.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'new_customer',
       'product_id', 'add_to_cart_order', 'reordered', '_merge',
       'product_name', 'aisle_id', 'department_id', 'prices',
       'price_range_loc', 'Busiest day', 'Busyness_Level', 'Busiest days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag', 'average_price',
       'spending_flag', 'ordering_behavior', 'order_frequency', 'Name',
       'Surname', 'Gender', 'State', 'Age', 'date_joined', 'n_dependants',
       'Income'],
      dtype='object')

## 5. Export Data Frame

In [37]:
# Export data to pkl

df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data ', 'orders_products_combined_customers_update_3.pkl'))